### PR TITLE
Fix #250: add ability to restrict ports

### LINKED ISSUE #250: add ability to restrict ports to certain ip ranges
Like
```
network:
   restrict_ports:
       https: [ 192.168.1.1/24 ]
```

### PR DESCRIPTION
**Pull request by e-carlin**
You can define
```
network:
  restrict_ports:
    https: [ 192.168.1.1/24 ]
```

That will restrict traffic on https port to traffic from only 192.168.1.1/24.

http will be added for you.

### CODE REVIEW

**Comment by e-carlin** [comment]
I just realized `iptables-restore v1.4.21: ! not allowed with multiple source or destination IP addresses` . Do we need to support a list of ips for the restrict or can it just be a single ip address CIDR range for now?

**Comment by e-carlin** [comment]
nvm it needs to be a list. fixing...

**Comment by e-carlin** [comment]
@robnagler ready for a review

**Comment by robnagler** [comment]
@e-carlin we can talk about a lot of changes. Review my changes.

File: tests/pkcli/build_data/1-fconf.out/srv/host/v9.radia.run/etc/sysconfig/iptables [hunk 1]
```diff
@@ -6,14 +6,12 @@
...
 -A INPUT -p icmp -j ACCEPT
-    -A INPUT -i em2 -p tcp -m state --state NEW -m tcp -s 192.168.1.1/24,127.0.0.1 --dport https -j ACCEPT
-    -A INPUT -i em2 -p tcp -m state --state NEW -m tcp --dport https -j DROP
-    -A INPUT -i em2 -p tcp -m state --state NEW -m tcp -s 192.168.1.1/24,127.0.0.1 --dport http -j ACCEPT
-    -A INPUT -i em2 -p tcp -m state --state NEW -m tcp --dport http -j DROP
     -A INPUT -i lo -j ACCEPT
     -A INPUT -i em1 -j ACCEPT
-    -A INPUT -i em2 -p tcp -m state --state NEW -m tcp --match multiport --dports 3100,3101,9999,http,https,pop3s,smtp,submission -j ACCEPT
 -A INPUT -i em2 -m state --state RELATED,ESTABLISHED -j ACCEPT
+    -A INPUT -i em2 -p tcp -m state --state NEW -m tcp --match multiport --dports 3100,3101,9999,http,pop3s,smtp,submission -j ACCEPT
```
Comment (e-carlin) [inline tests/pkcli/build_data/1-fconf.out/srv/host/v9.radia.run/etc/sysconfig/iptables:12]: This still has http. Are we assuming nginx always redirects http to https?

File: rsconf/component/network.py [hunk 2]
```diff
@@ -100,6 +100,7 @@ def internal_build_write(self):
         jc = self.j2_ctx
         jc = self.j2_ctx
         z = jc.network
+        self._restricted_public_tcp_ports(z)
```
Comment (e-carlin) [inline rsconf/component/network.py:103]: I think it is worth commenting this bit of code that it is important it is the last thing to mutate the lists of ports.

**Review by e-carlin — COMMENTED** [review]
Looks good. 

I still don't understand drop vs reject. You say that drop tells the attacker something. [This](https://serverfault.com/questions/157375/reject-vs-drop-when-using-iptables#:~:text=The%20difference%20is%20that%20the,need%20to%20wait%20for%20timeout.&text=DROP%20will%20cause%20packet%20to,cannot%20hide%20existence%20of%20server.) leads me to believe that reject tells them something and drop tells them nothing.

File: tests/pkcli/build_data/1-fconf.out/srv/host/v9.radia.run/etc/sysconfig/iptables (hunk 1 above)
Comment (robnagler) [reply tests/pkcli/build_data/1-fconf.out/srv/host/v9.radia.run/etc/sysconfig/iptables:12]: We don't allow anything over http except a redirect to https. For this case, we only need https blocked. This is simpler and clearer for now. Eventually we could support port ranges, but it's an unnecessary complication now.

**Comment by robnagler** [comment]
> I still don't understand drop vs reject. You say that drop tells the attacker something. This leads me to believe that reject tells them something and drop tells them nothing.

If you read the all the replies, you will see the debate, e.g. [this comment](https://serverfault.com/questions/157375/reject-vs-drop-when-using-iptables#comment1418172_157376):

> DROP will cause packet to timeout but it cannot hide existence of server. Try nmap -sS targethost sometime. It will be able to tell the difference between true closed and filtered ports based on SYN behavior.

You have to DROP everything on the host for it to be "invisible". If the host is in any way visible (ping response, for example), then a DROP tells you more than a REJECT.

File: rsconf/component/network.py (hunk 2 above)
Comment (robnagler) [reply rsconf/component/network.py:105]: c24d8fa added comments to iptables as well.

**Comment by e-carlin** [comment]
> If you read the all the replies, you will see the debate, e.g. [this comment](https://serverfault.com/questions/157375/reject-vs-drop-when-using-iptables#comment1418172_157376)

Interesting, thanks.

This comment further down was also illuminating:
> Without iptables rules, a closed port defaults to REJECT. If you DROP every port it's a fine alternative (your host appears down) but if you DROP only specific ports you're actually giving them more information than REJECT. If someone uses a blanket probe like nmap, specific ports that drop packets will appear as "filtered", which means they know you have a service there that you're hiding.

**Comment by robnagler** [comment]
That's the way I think about it. 

Anything that is on the Internet is visible in some way. Without that, you are doing to spend a lot of time debugging. If you care about these things, you use a firewall (also discussed in the thread).

**Pull request merged**